### PR TITLE
Feature: trade companion

### PR DESCRIPTION
**Pull request by WhiteFang5**
## Description

This PR adds a Trade Companion (TC) feature, based on MercuryTrade and some ground-works made in #107
The main feature of the Trade Companion include the following:
* Stash grid overlay for both Normal & Quad tabs
* Trade notification panel that can be positioned anywhere on any screen
* Trade notifications for both incoming and outgoing English trade and bulk-exchange whispers from the official poe website
* Customizable buttons, messages and actions for the trade notifications
  
From a tech p.o.v. the following features also got added/changes:
* Multi monitor support to allow moving the TC outside of the PoE client bounds
* A client.txt file watcher
* A drag & resize directive (to allow moving/resizing of the TC panel and grid stash)
* Shortcuts now behave like a queue
  
### What's next?
During a closed beta a number of bugs were fixed and a number of features were proposed as well.  
With this PR being the public beta, the aim is to find more bugs and get them fixed for a stable release.  
From that point on we can start adding more features on top of the TC.  
Here is an unsorted list of feature requests I've received/collected so far:
* ~(WhiteFang) Sound Effects~ Added to this PR in 02358465e4b1f273c637c3a8caad2633eb492836
* ~(WhiteFang) Color schemes~ Added to this PR in ff2ffde
* (WhiteFang) Add HO button
* (WhiteFang) Always On Top (TC only)
* (WhiteFang) Support trade messages from other trade sites
* ~(WhiteFang) Support trade messages in other languages~ Added to this PR in f98d4db
* ~(WhiteFang) Support for item sizes when highlighting items~ Added to this PR in 4ed993f
* (WhiteFang) Support leaving of party (requires character name -> maybe use oauth and allow choosing in the settings?)
* ~(FutureFear) Choose notification direction (top -> bottom OR bottom -> top)~ Added to this PR in 8059e28
* ~(Wratho) Transparent BG but NOT text~ Added to this PR in cbe088d
* (Razunter) Trade companion feature request: text placeholders in messages, like {item} name or {price}. Currently SlimTrade is the best trade tool of this kind for me, not Mercury.
* (Wratho) Support other languages
  * This mainly is for people who speak the supported languages and can translate the newly added TC section in english.json

## How Has This Been Tested?

- [ ] ran `npm run ng:lint`
- [ ] ran `npm run format`
- [ ] ran `npm run ng:test`
- [ ] added unit tests
- [x] manually tested

### CODE REVIEW

File: src/app/core/service/dialog/dialog-ref.service.ts [hunk 1]
```diff
@@ -29,33 +31,46 @@ export class DialogRefService {
...
-        VisibleFlag.Game | VisibleFlag.Dialog,
-        VisibleFlag.Overlay | VisibleFlag.Dialog
-      )
-      .subscribe(() => this.closeAll())
-  }
+    if (!this.escapeSubscription) {
+      const clearShortcut = () => {
+        this.escapeSubscription?.unsubscribe()
+        this.escapeSubscription = null
+      }
```
Comment (wratho) [inline src/app/core/service/dialog/dialog-ref.service.ts:38]: this seems weird. you are only clearing the subscription if the subscription is falsey?

File: src/app/core/service/input/shortcut.service.ts [hunk 2]
```diff
@@ -6,9 +6,16 @@ import { Observable, Subject } from 'rxjs'
 
 export interface Shortcut {
   accelerator: string
+  ref: any
```
Comment (wratho) [inline src/app/core/service/input/shortcut.service.ts:9]: i recommend `unknown` over `any` it forces you to cast as type when you get to a type inferrance. feel free to ignore it tho

File: src/app/layout/page/user-settings/user-settings.component.scss [hunk 3]
```diff
@@ -1,12 +1,11 @@
 ::ng-deep {
   .container-after-titlebar {
     background: #1c1c1c;
+    bottom: 36px !important;
```
Comment (wratho) [inline src/app/layout/page/user-settings/user-settings.component.scss:4]: ideally you avoid `!important`. this has a chance of making you sad later. If this is unavoidable its fine, but you should try using more specificity in your css selectors instead.

File: src/app/modules/trade-companion/component/trade-notifications-panel/trade-notifications-panel.component.scss [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      margin-left: auto;
+      margin-right: 0px;
+    }
+
+    .movable {
+      cursor: move;
+    }
+  }
+
+  .notifications {
```
Comment (wratho) [inline src/app/modules/trade-companion/component/trade-notifications-panel/trade-notifications-panel.component.scss:26]: probably don't want empty selectors

File: src/app/modules/trade-companion/component/trade-notifications-panel/trade-notifications-panel.component.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    windowBounds.y = offset
+    windowBounds.width -= offset * 2
+    windowBounds.height -= offset * 2
+
+    if (this.intersects(bounds, windowBounds)) {
+      this.boundsUpdate$.next(bounds)
+    }
+  }
+
+  public toggleGrid(): void {
```
Comment (wratho) [inline src/app/modules/trade-companion/component/trade-notifications-panel/trade-notifications-panel.component.ts:90]: empty function here

File: src/app/modules/trade-companion/trade-companion.module.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  public getFeatures(settings: TradeCompanionUserSettings): Feature[] {
+    return [{
+      name: 'stash-grid',
+      accelerator: undefined,
+    }]
+  }
+
+  public run(feature: string, settings: TradeCompanionUserSettings): void {
```
Comment (wratho) [inline src/app/modules/trade-companion/trade-companion.module.ts:95]: another empty function

File: src/app/shared/module/material/directive/resize-drag.directive.ts [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+      x: event.clientX,
+      y: event.clientY,
+    }
+    if (!this.overlaps(this.element, point)) {
+      return
+    }
+
+    this.mouseDownPosition = point
+    this.mouseDownBounds = { ...this.bounds }
+    /*const elementBounds = this.element.getBoundingClientRect()
```
Comment (wratho) [inline src/app/shared/module/material/directive/resize-drag.directive.ts:228]: do you want to keep these comments around?

**Review by wratho — CHANGES_REQUESTED** [review]

File: src/app/core/service/input/shortcut.service.ts (hunk 2 above)
Comment (WhiteFang5) [reply src/app/core/service/input/shortcut.service.ts:9]: I've seen "any" being used in a number of places, so I'll just stick with it.

File: src/app/layout/page/user-settings/user-settings.component.scss (hunk 3 above)
Comment (WhiteFang5) [reply src/app/layout/page/user-settings/user-settings.component.scss:4]: I've looked for a number of solutions, but without it, it doesn't seem to work, so I'm going to keep it for now.

File: src/app/modules/trade-companion/trade-companion.module.ts (hunk 6 above)
Comment (WhiteFang5) [reply src/app/modules/trade-companion/trade-companion.module.ts:93]: Has to be around (due to interface implementation), so I've added a comment to it. (commit: 5b14e87)

File: src/app/modules/trade-companion/component/trade-notifications-panel/trade-notifications-panel.component.ts (hunk 5 above)
Comment (WhiteFang5) [reply src/app/modules/trade-companion/component/trade-notifications-panel/trade-notifications-panel.component.ts]: Removed in 5b14e87

File: src/app/modules/trade-companion/component/trade-notifications-panel/trade-notifications-panel.component.scss (hunk 4 above)
Comment (WhiteFang5) [reply src/app/modules/trade-companion/component/trade-notifications-panel/trade-notifications-panel.component.scss]: Removed in 5b14e87

File: src/app/core/service/dialog/dialog-ref.service.ts (hunk 1 above)
Comment (WhiteFang5) [reply src/app/core/service/dialog/dialog-ref.service.ts:38]: The `clearShortcut` method that's defined here, is passed on (to another method) that'll call it some time in the future. At which point this (or any other class that has access) could have already cleared the shortcut manually, that's why the check with `?` is inside the method.  
This entire if-block is here to only create/register the shotcut once, regardless of how many times the method it resides in is called.

**Pull request merged**